### DISCUSSION TITLE
Delete and Sync related issues

### DISCUSSION THREAD

**Discussion by super-salad** [massif-press/compcon#2124] [Archive]
There are quite a few delete and sync related issues coming in over the Discord channel, and submitted here on the repo. Would it be possible to create a tag or a milestone to start organizing all of these so we can get a picture of the problem and provide some context so we can start working on it?

**Comment by msprijatelj** [comment]
Created a "Data Management" milestone, though the "COMP/CON API" Milestone seemed to exist previously and covered some of these deletion/data management issues. @jarena3 If COMP/CON API is a more appropriate milestone, feel free to delete the one I created and move its issues to COMP/CON API!

**Comment by super-salad** [comment]
This is great @msprijatelj! I will start taking a look to see if there's anything that jumps out at me that I might be able to help with, but I'm a bit wary of touching any of the backend stuff just yet. But no time like the present to start learning.

**Comment by jarena3** [comment]
yes please, any help would be greatly appreciated. I thought I found the culprit like three or four times now but these issues still persist

**Reply by super-salad** [reply]
Which files have you been working on in the past? It would be helpful to have a starting point. Right now I'm not sure if it's the timing of calling the sync commands, the commands themselves, or maybe the GraphQL queries?

**Reply by msprijatelj** [reply]
First thing I'd check is how the local and cloud save dates are being compared. I'm looking at an exported pilot right now and seeing dates saved as strings:
```json
    "lastModified": "Mon Sep 19 2022 16:31:20 GMT-0400 (Eastern Daylight Time)",
    "isDeleted": false,
    "expireTime": "",
    "deleteTime": "",
    "lastUpdate_cloud": "Mon Sep 19 2022 16:58:01 GMT-0400 (Eastern Daylight Time)",
    "lastSync": "Mon Sep 19 2022 16:58:01 GMT-0400 (Eastern Daylight Time)",
    "shareCodeExpiry": "Tue Dec 06 2022 18:40:06 GMT-0500 (Eastern Standard Time)",
```
My gut reaction is that if these strings are getting compared, they may get gunked up with this format unless the strings are explicitly converted into integer timestamps.  E.g. "Mon" always looking "greater" than "Tue" or something.

**Reply by Hawkzed** [reply]
> First thing I'd check is how the local and cloud save dates are being compared. I'm looking at an exported pilot right now and seeing dates saved as strings:
> 
> ```json
>     "lastModified": "Mon Sep 19 2022 16:31:20 GMT-0400 (Eastern Daylight Time)",
>     "isDeleted": false,
>     "expireTime": "",
>     "deleteTime": "",
>     "lastUpdate_cloud": "Mon Sep 19 2022 16:58:01 GMT-0400 (Eastern Daylight Time)",
>     "lastSync": "Mon Sep 19 2022 16:58:01 GMT-0400 (Eastern Daylight Time)",
>     "shareCodeExpiry": "Tue Dec 06 2022 18:40:06 GMT-0500 (Eastern Standard Time)",
> ```
> 
> My gut reaction is that if these strings are getting compared, they may get gunked up with this format unless the strings are explicitly converted into integer timestamps. E.g. "Mon" always looking "greater" than "Tue" or something.

On this note I've been looking at what gets exported pilot data and noticed that the "lastModified" value doesn't seem to change in accordance with local changes.

I can't seem to map it directly to cloud save functions and the like so perhaps it's of no consequence.

This is the same file.
(Images used to show changes)
![image](https://user-images.githubusercontent.com/15632927/206010855-8d1d0be9-5b7e-4392-9e27-4b00e81018a4.png)
![image](https://user-images.githubusercontent.com/15632927/206011731-2a456b7f-2f16-4163-b82c-17eee7d41601.png)

 I generated these results by:
1. Creating a character then.
2. Using the home button
3. Accessing the pilot roster
4. Exporting the new pilot to clipboard
5. Editing the character
6. Using the home button
3. Accessing the pilot roster
7. Exporting the modified pilot to clipboard

_______________________

On another note Javascript's default Date objects are known to be a bit crap. The .parse() function especially (some may call it unusable [[1]](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) links to -> [[2]](https://tc39.es/proposal-temporal/docs/index.html) links to -> [[3 - Section: Broken Parser – A Web Reality Issue]](https://maggiepint.com/2017/04/11/fixing-javascript-date-web-compatibility-and-reality/), [[4 - Also has a variety of examples]](https://stackoverflow.com/questions/5619202/parsing-a-string-to-a-date-in-javascript)) and it is used in some of the save comparisons:

```
function determineLatest(cloudItem: CollectionItem, localItem: CollectionItem): string {
  if (cloudItem.lastModifiedCloud === localItem.lastModifiedLocal) return 'synced'
  if (Date.parse(cloudItem.lastModifiedCloud) > Date.parse(localItem.lastModifiedLocal)) {
    return 'cloud'
  } else {
    return 'local'
  }
}
```

It could be worth integrating a well known replacement such as [Luxon](https://moment.github.io/luxon/#/) to maybe prevent issues in the future if it ever presents a problem.

**Reply by Hawkzed** [reply]
Having done some more investigation it appears that the 'LastModified' property is sometime out of date despite local changes having occurred leading to the sync comparing the correct last updated cloud key with the incorrect local key after the stale value is collected by the ProcessItemsList function.

Not that it's much evidence but after updating a character's description I excited to the main menu and these values were compared to determine the freshness of the local data, the cloud data was preferred despite local changes and my pilot reverted to the cloud version.
![image](https://user-images.githubusercontent.com/15632927/206288164-c2b3cb69-e060-4287-b23e-de5db5ae288b.png)

EDIT:
The Last Modified key updates when saving via the SaveController.ts but this change isn't reflected in item_sync.ts
![tempsnip](https://user-images.githubusercontent.com/15632927/206292782-25d278b9-fed3-4c96-a41e-694e9b319b70.png)

**Reply by Hawkzed** [reply]
Think I've fixed the problem along with @ptlm500.

Initially we diagnosed the problem with cloud syncing but also noticed a problem also occurred with local changes not persisting across browser refreshes. For example changing a pilot's name on refresh (with auto-sync off) forced the name to return to it's previous version.

It is related to the save() function in `src\classes\components\save\SaveController.ts`
Removing the ` || !this._isLoaded` from the if statement on line 33 makes the cloud save seem to be functional as well as local changes.

The save function was being bypassed by this early return after the vast majority of changes we tested in the UI.
Certain changes did persist on changes to the pilot profile such as level ups.

```
public save(skip?: boolean) {
    if (skip) return  // remove || !this._isLoaded
    this.IsDirty = true
    this.LastModified = new Date().toString()
    store.dispatch(`set_${this.Parent.ItemType}_dirty`)
    // const sp = getModule(UserStore, store).UserProfile.IsSavePerformant
    // if (!sp) {
    SaveAllLocalUpdates()
    // }
  }
```

Not sure of the knock-on effects but when testing locally it appears to have fixed the local LastModified value as well as fixing the issues we observed with saves not persisting locally.

Happy to create a PR, let me know what you find as I'm looking forward to using this tool with my players again.
